File: bindings/python/tests/test_index.py
"""Declarative indexes IDX.* (contract §3.8, §6). Remote-only."""

import time

import kevy
from kevy import IdxType


def _wait_ready(c, name, timeout=5.0):
    deadline = time.time() + timeout
    while time.time() < deadline:
        for info in c.idx_list():
            if info.name == name.encode() and info.state == "ready":
                return
        time.sleep(0.02)
    raise AssertionError(f"index {name} never became ready")


def test_range_paging_and_eq(remote_server):
    c = kevy.connect(remote_server.url)
    c.idx_create_range("byage", "user:", "age", IdxType.I64)
    for i, age in enumerate(["21", "22", "23", "24", "25"]):
        c.hset(f"user:{chr(ord('a') + i)}", "age", age)
    _wait_ready(c, "byage")

    infos = c.idx_list()
    assert any(i.name == b"byage" and i.kind == "range" for i in infos)  # unknown labels skipped

    # LIMIT 2 pages through 5 rows; cursor advances, ends at None.
    seen = 0
    cursor = None
    while True:
        page = c.idx_query_range("byage", "0", "100", 2, cursor)
        seen += len(page.rows)
        if page.cursor is None:
            break
        cursor = page.cursor
        assert seen <= 10
    assert seen == 5

    page = c.idx_query_eq("byage", "23", 10)
    assert len(page.rows) == 1 and page.rows[0].value == b"23"

    assert c.idx_drop("byage") is True
    assert c.idx_drop("byage") is False
    c.close()


def test_raw_query_passthrough(remote_server):
    c = kevy.connect(remote_server.url)
    c.idx_create_range("byid", "item:", "id", IdxType.I64)
    c.hset("item:1", "id", "1")
    _wait_ready(c, "byid")
    # raw COUNT via the escape hatch reaches server capability directly
    r = c.do(b"IDX.COUNT", b"byid", b"RANGE", b"0", b"100")
    assert r.kind.name in ("INT", "SIMPLE", "BULK")
    c.close()


def test_embedded_unsupported():
    c = kevy.connect("mem://idx-bus")
    for call in (
        lambda: c.idx_list(),
        lambda: c.idx_create_range("i", "u:", "age", IdxType.I64),
        lambda: c.idx_query_range("i", "0", "1", 10),
    ):
        try:
            call()
            assert False, "expected UnsupportedError"
        except kevy.UnsupportedError:
            pass
    c.close()

File: bindings/cpp/test/test_index.cpp
// Declarative indexes IDX.* — remote-only (contract §6 "IDX query").
#include <chrono>
#include <thread>

#include "harness.hpp"
#include "kevy/client.hpp"

using namespace kevy;
using namespace kevy::test;
using namespace std::chrono_literals;

static bool wait_idx_ready(Client& c, const std::string& name) {
  auto deadline = std::chrono::steady_clock::now() + std::chrono::seconds(5);
  while (std::chrono::steady_clock::now() < deadline) {
    for (const auto& in : c.idx_list())
      if (in.name == name && in.state == "ready") return true;
    std::this_thread::sleep_for(20ms);
  }
  return false;
}

KEVY_TEST(index_range_paging_and_eq) {
  if (!server_available()) return;
  auto srv = spawn_server();
  Client c = Client::connect(srv->url());

  c.idx_create_range("byage", "user:", "age", IdxType::I64);
  const char* ages[] = {"21", "22", "23", "24", "25"};
  for (int i = 0; i < 5; i++) {
    std::string key = std::string("user:") + char('a' + i);
    c.hset(key, {{"age", ages[i]}});
  }
  CHECK(wait_idx_ready(c, "byage"));

  // idx_list parses IdxInfo including kind.
  bool found = false;
  for (const auto& in : c.idx_list())
    if (in.name == "byage") {
      found = true;
      CHECK_EQ(in.kind, std::string("range"));
    }
  CHECK(found);

  // Range paging: LIMIT 2 pages through 5 rows, cursor ends at null.
  int seen = 0;
  std::optional<std::string> cursor;
  for (int guard = 0; guard < 10; guard++) {
    std::optional<std::string_view> cur;
    if (cursor.has_value()) cur = *cursor;
    IdxPage page = c.idx_query_range("byage", "0", "100", 2, cur);
    seen += static_cast<int>(page.rows.size());
    if (!page.cursor.has_value()) break;
    cursor = page.cursor;
  }
  CHECK_EQ(seen, 5);

  // EQ point lookup.
  IdxPage eq = c.idx_query_eq("byage", "23", 10);
  CHECK_EQ(eq.rows.size(), size_t(1));
  CHECK_EQ(eq.rows[0].value, std::string("23"));

  // drop reports existed, then gone.
  CHECK(c.idx_drop("byage"));
  CHECK(!c.idx_drop("byage"));
}

KEVY_TEST(index_raw_passthrough) {
  if (!server_available()) return;
  auto srv = spawn_server();
  Client c = Client::connect(srv->url());
  // The *_raw escape hatch reaches the full grammar (here: COUNT via command).
  c.idx_create_range("bynum", "n:", "v", IdxType::I64);
  c.hset("n:1", {{"v", "10"}});
  wait_idx_ready(c, "bynum");
  Reply r = c.idx_query_raw({"bynum", "RANGE", "0", "100"});
  CHECK(r.kind == ReplyKind::Array);
}
